ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,264 @@
+
+The proposed name for this contract is `ValidatorSetManager`.
+
+This proposal should also help define the payload of the P-Chain notifications sent to the Subnet upon validator addition, weight change and removal.
+
+This ACP relies on concepts introduced in [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets). It depends on it to be marked as `Implementable`.
+
+## Motivation
+
+[ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets) opens the door to managing a Subnet validator set (stored on the P-Chain) from any chain on the Avalanche Network. The P-Chain allows a Subnet to specify a Warp Derived Address as the validator set manager. The `(blockchainID, address)` pair specified by the Warp Derived Address is responsible for sending Warp `AddressedCall` payloads to issue `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` on the P-Chain. This enables an onchain program add, modify the weight of, and remove validators.
+
+On each validator set change, the P-Chain generates a Warp message to notify any onchain program tracking the validator set. Onchain programs must be able to interpret this message, so they can trigger the appropriate action. These notification messages have not been specified to date.
+
+Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based, we propose defining a reference implementation for a Solidity smart contract that can:
+
+1. Hold relevant information about the current Subnet validator set, as well as historical information
+2. Send validator set updates to the P-Chain by generating `AdressedCall`s defined in ACP-77
+3. Correctly update the validator set by interpreting notification messages received from the P-Chain
+4. Be easily extended with custom “security modules” to implement any security model
+
+Having an audited and open-source reference implementation freely available will contribute to lowering the cost of launching Subnets on Avalanche.
+
+Once deployed, this contract will be used as the `address` in the `SetSubnetValidatorManagerTx`.
+
+## Specification
+
+### Storage
+
+The `ValidatorSetManager` needs to store all this data to operate correctly.
+
+**Note:** This is pseudo-code that might not be 100% syntactically correct.
+
+```solidity
+// The ID of the Subnet that uses this contract to manage its validator set
+bytes32 subnetID;
+
+// The current Subnet validator set (list of NodeIDs)
+bytes32[] subnetCurrentValidatorSet;
+
+// Subnet validator
+struct SubnetValidator {
+    // The Warp message ID corresponding to the RegisterSubnetValidatorTx that added the validator.
+    // The message ID is needed to submit subsequent SetSubnetValidatorWeightTx.
+    bytes32 messageID,
```

Comment:
think this ID should be a validation ID instead, denoting a hash of the info for a validation period when registering

Reply:
Absolutely. I am working on an updated version of this ACP reflecting the latest changes in ACP-77!